Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWindowsCocoa.mm

```diff
@@ -0,0 +1,296 @@
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+#import "config.h"
+#import "WebExtensionContext.h"
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+#import "CocoaHelpers.h"
+#import "WebExtensionContextProxy.h"
+#import "WebExtensionContextProxyMessages.h"
+#import "WebExtensionWindowIdentifier.h"
+#import "_WKWebExtensionControllerDelegatePrivate.h"
+#import "_WKWebExtensionWindowCreationOptionsInternal.h"
+
+namespace WebKit {
+
+static inline bool matchesFilter(const WebExtensionWindow& window, OptionSet<WebExtensionWindow::TypeFilter> filter)
+{
+    switch (window.type()) {
+    case WebExtensionWindow::Type::Normal:
+        return filter.contains(WebExtensionWindow::TypeFilter::Normal);
+
+    case WebExtensionWindow::Type::Popup:
+        return filter.contains(WebExtensionWindow::TypeFilter::Popup);
+    }
+}
+
+void WebExtensionContext::windowsCreate(WebExtensionWindowParameters creationParameters, CompletionHandler<void(std::optional<WebExtensionWindowParameters>, WebExtensionWindow::Error)>&& completionHandler)
+{
+    auto delegate = extensionController()->delegate();
+    if (![delegate respondsToSelector:@selector(webExtensionController:openNewWindowWithOptions:forExtensionContext:completionHandler:)]) {
+        completionHandler(std::nullopt, "windows.create() not implemented."_s);
+        return;
+    }
+
+    static constexpr CGFloat NaN = std::numeric_limits<CGFloat>::quiet_NaN();
+    static constexpr CGRect CGRectNaN = { { NaN, NaN }, { NaN, NaN } };
+
+    auto *creationOptions = [[_WKWebExtensionWindowCreationOptions alloc] _init];
+    creationOptions.desiredWindowType = creationParameters.type ? toAPI(creationParameters.type.value()) : _WKWebExtensionWindowTypeNormal;
+    creationOptions.desiredWindowState = creationParameters.state ? toAPI(creationParameters.state.value()) : _WKWebExtensionWindowStateNormal;
+    creationOptions.shouldFocus = creationParameters.focused && creationParameters.focused.value();
+    creationOptions.shouldUsePrivateBrowsing = creationParameters.privateBrowsing && creationParameters.privateBrowsing.value();
+
+    if (creationParameters.frame) {
+        CGRect desiredFrame = creationParameters.frame.value();
+
+#if PLATFORM(MAC)
+        // Window coordinates on macOS have the origin in the bottom-left corner.
+        // Web Extensions have window coordinates in the top-left corner.
+        CGRect screenFrame = NSScreen.mainScreen.frame;
```

Comment:
Will all windows be created on the main screen? Or would this be wrong if the browser was in the non-main screen?

Reply:
On macOS the other screens are rooted in the main screen coords. So an extension could still position on a second screen by using negative positions or positive positions wider than the main screen, depending on where the other screen is positioned. So using mainScreen here is really the only choice.